### PR TITLE
Simplify `CapabilitySet` Diagnostic Printing

### PR DESCRIPTION
**Pull request by ArielG-NV**
Fixes: #4675
Fixes: #4683
Fixes: #4443
Fixes: #4585
Fixes: #4172

Made the following changes:
1. All capability diagnostic printing logic tries to simplify before printing. This means that we do not print atoms which imply another atom.
2. Do not print the `_` prefix part of atom names since it is misleading users on what they should use to solve a capability issue encountered. (`_Internal` `External` atom changes are not in this PR)
3. Bundle together printing of all sets which contain exactly the same atoms (excluding abstract atoms). This allows printing the following `vertex/fragment/hull/domain/... + glsl` instead of `vertex + glsl | fragment + glsl | hull + glsl | domain + glsl | ....`
4. Rework how entry-point errors are reported to users (example at bottom of PR comment)
5. Rework how atom-provenance data is collected to be leaner and more useful so we can rework the errors. There are 2 notable changes here:
    * We no longer store a list which describes where the first of an `CapabilityAtom` comes from. This heavily simplifies AST logic for the capability system. AST parsing of capabilities is much faster. The trade-off is faster AST parsing and correct AST node data for slower diagnostics if an error is found
    * atom-provenance data now stores a reference to an atom's use-site to provide information on **where** and **what** is wrong with user code versus only sharing **what** and not where. 

---

case 1 of example output which needs to be addressed (test `capabilitySimplification1.slang`):
```
shader.hlsl(11): error 36107: entrypoint 'main' requires capability 'textualTarget + cuda + cuda_sm_1_0 + cuda_sm_2_0 + cuda_sm_3_0 + cuda_sm_3_5 + cuda_sm_4_0 + cuda_sm_5_0 + cuda_sm_6_0 + cuda_sm_7_0 + vertex | textualTarget + cuda + cuda_sm_1_0 + cuda_sm_2_0 + cuda_sm_3_0 + cuda_sm_3_5 + cuda_sm_4_0 + cuda_sm_5_0 + cuda_sm_6_0 + cuda_sm_7_0 + fragment | textualTarget + cuda + cuda_sm_1_0 + cuda_sm_2_0 + cuda_sm_3_0 + cuda_sm_3_5 + cuda_sm_4_0 + cuda_sm_5_0 + cuda_sm_6_0 + cuda_sm_7_0 + compute | textualTarget + cuda + cuda_sm_1_0 + cuda_sm_2_0 + cuda_sm_3_0 + cuda_sm_3_5 + cuda_sm_4_0 + cuda_sm_5_0 + cuda_sm_6_0 + cuda_sm_7_0 + hull | textualTarget + cuda + cuda_sm_1_0 + cuda_sm_2_0 + cuda_sm_3_0 + cuda_sm_3_5 + cuda_sm_4_0 + cuda_sm_5_0 + cuda_sm_6_0 + cuda_sm_7_0 + domain | textualTarget + cuda + cuda_sm_1_0 + cuda_sm_2_0 + cuda_sm_3_0 + cuda_sm_3_5 + cuda_sm_4_0 + cuda_sm_5_0 + cuda_sm_6_0 + cuda_sm_7_0 + geometry | textualTarget + cuda + cuda_sm_1_0 + cuda_sm_2_0 + cuda_sm_3_0 + cuda_sm_3_5 + cuda_sm_4_0 + cuda_sm_5_0 + cuda_sm_6_0 + cuda_sm_7_0 + raygen | textualTarget + cuda + cuda_sm_1_0 + cuda_sm_2_0 + cuda_sm_3_0 + cuda_sm_3_5 + cuda_sm_4_0 + cuda_sm_5_0 + cuda_sm_6_0 + cuda_sm_7_0 + intersection | textualTarget + cuda + cuda_sm_1_0 + cuda_sm_2_0 + cuda_sm_3_0 + cuda_sm_3_5 + cuda_sm_4_0 + cuda_sm_5_0 + cuda_sm_6_0 + cuda_sm_7_0 + anyhit | textualTarget + cuda + cuda_sm_1_0 + cuda_sm_2_0 + cuda_sm_3_0 + cuda_sm_3_5 + cuda_sm_4_0 + cuda_sm_5_0 + cuda_sm_6_0 + cuda_sm_7_0 + closesthit | textualTarget + cuda + cuda_sm_1_0 + cuda_sm_2_0 + cuda_sm_3_0 + cuda_sm_3_5 + cuda_sm_4_0 + cuda_sm_5_0 + cuda_sm_6_0 + cuda_sm_7_0 + miss | textualTarget + cuda + cuda_sm_1_0 + cuda_sm_2_0 + cuda_sm_3_0 + cuda_sm_3_5 + cuda_sm_4_0 + cuda_sm_5_0 + cuda_sm_6_0 + cuda_sm_7_0 + mesh | textualTarget + cuda + cuda_sm_1_0 + cuda_sm_2_0 + cuda_sm_3_0 + cuda_sm_3_5 + cuda_sm_4_0 + cuda_sm_5_0 + cuda_sm_6_0 + cuda_sm_7_0 + amplification | textualTarget + cuda + cuda_sm_1_0 + cuda_sm_2_0 + cuda_sm_3_0 + cuda_sm_3_5 + cuda_sm_4_0 + cuda_sm_5_0 + cuda_sm_6_0 + cuda_sm_7_0 + callable | textualTarget + hlsl + sm_4_0 + sm_4_1 + sm_5_0 + sm_5_1 + sm_6_0 + sm_6_1 + sm_6_2 + sm_6_3 + sm_6_4 + sm_6_5 + vertex | textualTarget + hlsl + sm_4_0 + sm_4_1 + sm_5_0 + sm_5_1 + sm_6_0 + sm_6_1 + sm_6_2 + sm_6_3 + sm_6_4 + sm_6_5 + fragment | textualTarget + hlsl + sm_4_0 + sm_4_1 + sm_5_0 + sm_5_1 + sm_6_0 + sm_6_1 + sm_6_2 + sm_6_3 + sm_6_4 + sm_6_5 + compute | textualTarget + hlsl + sm_4_0 + sm_4_1 + sm_5_0 + sm_5_1 + sm_6_0 + sm_6_1 + sm_6_2 + sm_6_3 + sm_6_4 + sm_6_5 + hull | textualTarget + hlsl + sm_4_0 + sm_4_1 + sm_5_0 + sm_5_1 + sm_6_0 + sm_6_1 + sm_6_2 + sm_6_3 + sm_6_4 + sm_6_5 + domain | textualTarget + hlsl + sm_4_0 + sm_4_1 + sm_5_0 + sm_5_1 + sm_6_0 + sm_6_1 + sm_6_2 + sm_6_3 + sm_6_4 + sm_6_5 + geometry | textualTarget + hlsl + sm_4_0 + sm_4_1 + sm_5_0 + sm_5_1 + sm_6_0 + sm_6_1 + sm_6_2 + sm_6_3 + sm_6_4 + sm_6_5 + raygen | textualTarget + hlsl + sm_4_0 + sm_4_1 + sm_5_0 + sm_5_1 + sm_6_0 + sm_6_1 + sm_6_2 + sm_6_3 + sm_6_4 + sm_6_5 + intersection | textualTarget + hlsl + sm_4_0 + sm_4_1 + sm_5_0 + sm_5_1 + sm_6_0 + sm_6_1 + sm_6_2 + sm_6_3 + sm_6_4 + sm_6_5 + anyhit | textualTarget + hlsl + sm_4_0 + sm_4_1 + sm_5_0 + sm_5_1 + sm_6_0 + sm_6_1 + sm_6_2 + sm_6_3 + sm_6_4 + sm_6_5 + closesthit | textualTarget + hlsl + sm_4_0 + sm_4_1 + sm_5_0 + sm_5_1 + sm_6_0 + sm_6_1 + sm_6_2 + sm_6_3 + sm_6_4 + sm_6_5 + miss | textualTarget + hlsl + sm_4_0 + sm_4_1 + sm_5_0 + sm_5_1 + sm_6_0 + sm_6_1 + sm_6_2 + sm_6_3 + sm_6_4 + sm_6_5 + mesh | textualTarget + hlsl + sm_4_0 + sm_4_1 + sm_5_0 + sm_5_1 + sm_6_0 + sm_6_1 + sm_6_2 + sm_6_3 + sm_6_4 + sm_6_5 + amplification | textualTarget + hlsl + sm_4_0 + sm_4_1 + sm_5_0 + sm_5_1 + sm_6_0 + sm_6_1 + sm_6_2 + sm_6_3 + sm_6_4 + sm_6_5 + callable', which is incompatible with the current compilation target 'spirv_1_0 + spirv_1_1 + spirv_1_2 + spirv_1_3 + spirv_1_4 + spirv_1_5 + fragment + SPV_KHR_fragment_shader_barycentric + SPV_EXT_fragment_fully_covered + SPV_EXT_shader_atomic_float_add + SPV_EXT_shader_atomic_float_min_max + SPV_EXT_mesh_shader + SPV_KHR_ray_tracing + SPV_KHR_ray_query + SPV_GOOGLE_user_type'.
uint4 main(int4 input0 : ATTR0, uint4 input1 : ATTR1) : SV_Target {
      ^~~~
hlsl.meta.slang(14867): note: see using of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see definition of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see using of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see definition of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see using of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see definition of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see using of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see definition of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see using of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see definition of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see using of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see definition of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see using of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see definition of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see using of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see definition of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see using of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see definition of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see using of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see definition of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see using of 'WaveMultiPrefixProduct'
hlsl.meta.slang(14867): note: see definition of 'WaveMultiPrefixProduct'
```

case 1 now:
```
7: entrypoint 'computeMain' does not support compilation target 'glsl' with stage 'compute'
void computeMain()
     ^~~~~~~~~~~
tests\language-feature\capability\capabilitySimplification3.slang(15): note: see using of 'WaveMultiPrefixProduct'
    WaveMultiPrefixProduct(1, 0);
    ^~~~~~~~~~~~~~~~~~~~~~
hlsl.meta.slang(15101): note: see definition of 'WaveMultiPrefixProduct'
T WaveMultiPrefixProduct(T value, uint4 mask)
  ^~~~~~~~~~~~~~~~~~~~~~
hlsl.meta.slang(15100): note: see declaration of 'require'
[require(cuda_hlsl, waveprefix)]
 ^~~~~~~
```

---

case 2 of what needs to be addressed (test `capability7.slang`):
```
tests/language-feature/capability/capability7.slang(26): error 36104: 'processDataBad' uses undeclared capability 'sm_5_0'.
void processDataBad()
     ^~~~~~~~~~~~~~
tests/language-feature/capability/capability7.slang(28): note: see using of 'processDataBadNested'
    processDataBadNested();
    ^~~~~~~~~~~~~~~~~~~~
tests/language-feature/capability/capability7.slang(20): note: see definition of 'processDataBadNested'
void processDataBadNested()
     ^~~~~~~~~~~~~~~~~~~~
tests/language-feature/capability/capability7.slang(26): error 36104: 'processDataBad' uses undeclared capability 'sm_5_1'.
void processDataBad()
     ^~~~~~~~~~~~~~
tests/language-feature/capability/capability7.slang(28): note: see using of 'processDataBadNested'
    processDataBadNested();
    ^~~~~~~~~~~~~~~~~~~~
tests/language-feature/capability/capability7.slang(20): note: see definition of 'processDataBadNested'
void processDataBadNested()
     ^~~~~~~~~~~~~~~~~~~~
...
```

case 2 now:
```
tests/language-feature/capability/capability7.slang(26): error 36104: 'processDataBad' uses undeclared capability 'sm_6_0'.
void processDataBad()
     ^~~~~~~~~~~~~~
tests/language-feature/capability/capability7.slang(28): note: see using of 'processDataBadNested'
    processDataBadNested();
    ^~~~~~~~~~~~~~~~~~~~
tests/language-feature/capability/capability7.slang(20): note: see definition of 'processDataBadNested'
void processDataBadNested()
     ^~~~~~~~~~~~~~~~~~~~
``` 

---

case 3 of what needs to be addressed (test `capabilitySimplification2.slang`):
```
tests\language-feature\capability\capabilitySimplification2.slang(25): warning 41012: entry point 'computeMain' uses additional capabilities that are not part of the specified profile 'sm_5_0'. The profile setting is automatically updated to include these capabilities: 'glsl + GLSL_130 + GLSL_140 + GLSL_330 + GLSL_400 + GLSL_430  + GLSL_450 + GL_EXT_nonuniform_qualifier + GL_KHR_shader_subgroup_arithmetic + GL_KHR_shader_subgroup_ballot + GL_KHR_shader_subgroup_quad + GL_KHR_shader_subgroup_shuffle + GL_KHR_shader_subgroup_vote'
```

case 3 now
```
tests\language-feature\capability\capabilitySimplification2.slang(25): warning 41012: entry point 'computeMain' uses additional capabilities that are not part of the specified profile 'sm_5_0'. The profile setting is automatically updated to include these capabilities: 'GLSL_450 + GL_EXT_nonuniform_qualifier + GL_KHR_shader_subgroup_arithmetic + GL_KHR_shader_subgroup_ballot + GL_KHR_shader_subgroup_quad + GL_KHR_shader_subgroup_shuffle + GL_KHR_shader_subgroup_vote'
```

### CODE REVIEW

**Comment by csyonghe** [comment]
I think there is another opportunity for simplification: when complaining that two caoapbility sets are incompatible, we can just print the "key" atoms in both sets instead of everything. Because only the key atoms are causing the incompatiblity.

**Comment by ArielG-NV** [comment]
> 
I think there is another opportunity for simplification: when complaining that two capability sets are incompatible, we can just print the "key" atoms in both sets instead of everything. Because only the key atoms are causing the incompatibility.

100% agree. 

I also think I will add a method to print sets (where the only difference is a shader stage) in the form of "vertex/fragment {capabilities}". This would be quite useful since normally capabilities overlap across stages

**Review by jkwak-work — COMMENTED** [review]
Looks good to me.

**Review by csyonghe — DISMISSED** [review]

File: source/slang/slang-diagnostic-defs.h [hunk 1]
```diff
@@ -737,8 +737,8 @@ DIAGNOSTIC(41001, Error, recursiveType, "type '$0' contains cyclic reference to
 
 DIAGNOSTIC(41010, Warning, missingReturn, "control flow may reach end of non-'void' function")
 DIAGNOSTIC(41011, Error, profileIncompatibleWithTargetSwitch, "__target_switch has no compatable target with current profile '$0'")
-DIAGNOSTIC(41012, Warning, profileImplicitlyUpgraded, "user set `profile` had an implicit upgrade applied to it, atoms added: '$0'")
-DIAGNOSTIC(41012, Error, profileImplicitlyUpgradedRestrictive, "user set `profile` had an implicit upgrade applied to it, atoms added: '$0'")
+DIAGNOSTIC(41012, Warning, profileImplicitlyUpgraded, "user set profile $0 had an implicit upgrade applied to it, atoms added: '$1'")
```
Comment (csyonghe) [inline source/slang/slang-diagnostic-defs.h:740]: "shader '$0' uses additional capabilities that are not part of the specified profile '$1'. The profile setting is automatically updated to include these capabilities: '$2'"

File: source/slang/slang-diagnostic-defs.h [hunk 2]
```diff
@@ -737,8 +737,8 @@ DIAGNOSTIC(41001, Error, recursiveType, "type '$0' contains cyclic reference to
...
+DIAGNOSTIC(41012, Error, profileImplicitlyUpgradedRestrictive, "user set profile $0 had an implicit upgrade applied to it, atoms added: '$1'")
```
Comment (csyonghe) [inline source/slang/slang-diagnostic-defs.h:741]: "shader '$0' uses capabilities that are not part of the specified profile '$1'. Missing capabilities are: '$2'"

**Comment by ArielG-NV** [comment]
> But the new message is still not clear to me of what is wrong.

This is a problem I should fix in this PR, otherwise this PR has no value. To fix clarity of error/warning diagnosis of entry-point incompatibility I plan to-do the following:
1. tell user the failing target+stage combo
2. diagnose based on stored capability providence (have to change how we store atom-providence) the exact function failing and where its definition is

**Comment by jkwak-work** [comment]
> th stage 'comput

**Comment by jkwak-work** [comment]
Sorry. I clicked a wrong button and accidently closed it; but reopened it right away.

**Comment by jkwak-work** [comment]
> ```
> tests/language-feature/capability/capabilitySimplification3.slang(32): error 36107: entrypoint 'computeMain' does not support compilation target 'glsl' with stage 'compute'.
> tests/language-feature/capability/capabilitySimplification3.slang(22): note: see using of 'WaveMultiPrefixProduct'
>     WaveMultiPrefixProduct(1, 0);
>     ^~~~~~~~~~~~~~~~~~~~~~
> hlsl.meta.slang(15101): note: see definition of 'WaveMultiPrefixProduct'
> ```

This error message looks great.
But the "note" should also show one line above where "require" is described.

**Review by csyonghe — APPROVED** [review]

**Pull request merged**